src/Cli/dotnet/commands/dotnet-sln/remove/Program.cs

```diff
@@ -2,58 +2,135 @@
+                            : fullPath);
+                });
+                RemoveProjectsAsync(solutionFileFullPath, relativeProjectPaths, CancellationToken.None).Wait();
+                return 0;
+            }
+            catch (Exception ex) when (ex is not GracefulException)
+            {
+                if (ex is SolutionException || ex.InnerException is SolutionException)
+                {
+                    throw new GracefulException(CommonLocalizableStrings.InvalidSolutionFormatString, solutionFileFullPath, ex.Message);
+                }
+                // TODO: Check
+                if (ex.InnerException is GracefulException)
+                {
+                    throw ex.InnerException;
+                }
+                throw new GracefulException(ex.Message, ex);
+            }
+        }
+
+        private async Task RemoveProjectsAsync(string solutionFileFullPath, IEnumerable<string> projectPaths, CancellationToken cancellationToken)
+        {
+            ISolutionSerializer serializer = SlnCommandParser.GetSolutionSerializer(solutionFileFullPath);
+            SolutionModel solution = await serializer.OpenAsync(solutionFileFullPath, cancellationToken);
 
-            bool slnChanged = false;
-            foreach (var path in relativeProjectPaths)
+            // set UTF8 BOM encoding for .sln
+            if (serializer is ISolutionSerializer<SlnV12SerializerSettings> v12Serializer)
             {
-                slnChanged |= slnFile.RemoveProject(path);
+                solution.SerializerExtension = v12Serializer.CreateModelExtension(new()
+                {
+                    Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: true)
+                });
             }
 
-            slnFile.RemoveEmptyConfigurationSections();
+            foreach (var projectPath in projectPaths)
+            {
+                var project = solution.FindProject(projectPath);
+                if (project != null)
+                {
+                    solution.RemoveProject(project);
+                    Reporter.Output.WriteLine(CommonLocalizableStrings.ProjectRemovedFromTheSolution, projectPath);
+                }
+                else
+                {
+                    Reporter.Output.WriteLine(CommonLocalizableStrings.ProjectNotFoundInTheSolution, projectPath);
+                }
+            }
 
-            slnFile.RemoveEmptySolutionFolders();
+            Dictionary<SolutionFolderModel, int> nonFolderDescendantsCount = new();
+            foreach (var item in solution.SolutionFolders)
+            {
+                CountNonFolderDescendants(solution, item, nonFolderDescendantsCount);
+            }
 
-            if (slnChanged)
+            var emptyFolders = nonFolderDescendantsCount.Where(i => i.Value == 0);
+            foreach (var folder in emptyFolders)
             {
-                slnFile.Write();
+                solution.RemoveFolder(folder.Key);
             }
```

Comment:
SolutionPersistence's `SolutionFolderModel` does not currently have a way of identifying children

Reply:
No. This is a different use case. https://github.com/microsoft/vs-solutionpersistence/pull/81 removes all children when removing a folder. The use case here is the other way around: the project gets removed, and then the solution is analyzed for empty folder structures that do not contain files or projects

Reply:
Eg, after removing `foo/bar/abcd/efgh/efgh.csproj`, all parent folders should be removed assuming they have no other children